GVFS/GVFS.Common/NuGetUpgrader/NuGetFeed.cs

```diff
@@ -0,0 +1,169 @@
+﻿using GVFS.Common.Tracing;
+using NuGet.Common;
+using NuGet.Packaging.Core;
+using NuGet.Protocol;
+using NuGet.Protocol.Core.Types;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace GVFS.Common.NuGetUpgrader
+{
+    /// <summary>
+    /// Handles interactions with a NuGet Feed.
+    /// </summary>
+    public class NuGetFeed
+    {
+        private ITracer tracer;
+        private string feedUrl;
+        private string feedName;
+        private string downloadFolder;
+        private string personalAccessToken;
+
+        public NuGetFeed(
+            string feedUrl,
+            string feedName,
+            string downloadFolder,
+            string personalAccessToken,
+            ITracer tracer)
+        {
+            this.feedUrl = feedUrl;
+            this.feedName = feedName;
+            this.downloadFolder = downloadFolder;
+            this.personalAccessToken = personalAccessToken;
+            this.tracer = tracer;
+        }
+
+        private NuGet.Configuration.PackageSourceCredential Credentials
+        {
+            get
+            {
+                return NuGet.Configuration.PackageSourceCredential.FromUserInput(
+                    "VfsForGitNugetUpgrader",
+                    "PersonalAccessToken",
+                    this.personalAccessToken,
+                    false);
+            }
+        }
+
+        /// <summary>
+        /// Query a NuGet feed for list of packages that match the packageId.
+        /// </summary>
+        /// <param name="packageId"></param>
+        /// <returns>List of packages that match query parameters</returns>
+        public virtual async Task<IList<IPackageSearchMetadata>> QueryFeed(string packageId)
+        {
+            SourceRepository sourceRepository = Repository.Factory.GetCoreV3(this.feedUrl);
+            if (!string.IsNullOrEmpty(this.personalAccessToken))
+            {
+                sourceRepository.PackageSource.Credentials = this.Credentials;
+            }
+
+            var packageMetadataResource = await sourceRepository.GetResourceAsync<PackageMetadataResource>();

```

Comment:
Out of curiosity, is there a non-async version of `GetResource`?

Reply:
There are some non-async versions of these methods, but not all. Looking at the code, the non-async versions seem to call through to the async versions. I think `GetResource` might have a non-async version, but `GetDownloadResulAsync` does not, for instance.